sbol2/componentdefinition.py

```diff
@@ -164,29 +165,102 @@ def addRole(self, new_role):
     def removeRole(self, index=0):
         self._roles.remove(index)
 
-    def assemble(self, components, assembly_standard="", doc=None):
+    def assemble(self, component_list, assembly_method=None, doc=None):
         """Assembles ComponentDefinitions into an abstraction hierarchy.
 
         The resulting data structure is a partial design,
         still lacking a primary structure or explicit sequence.
         To form a primary structure out of the ComponentDefinitions,
         call linearize after calling assemble.
         To fully realize the target sequence, use Sequence::assemble().
-        :param components: Either a list of URIs
+        :param component_list: Either a list of URIs
         for the constituent ComponentDefinitions or a list of subcomponents.
         A list of displayIds is also acceptable if using SBOL-compliant URIs.
-        :param assembly_standard: An optional argument
+        :param assembly_method: An optional argument
         such as IGEM_STANDARD_ASSEMBLY that affects how components
         are composed and the final target sequence.
         :param doc: The Document to which the assembled ComponentDefinitions
         will be added. If not set, then you must add this ComponentDefinition
         to a Document before calling this method.
         :return: None
         """
-        raise NotImplementedError("Not yet implemented")
 
-    def assemblePrimaryStructure(self, primary_structure,
-                                 assembly_standard="", doc=None):
+        # Due to the recursive nature of this routine, it is hard to completely
+        # validate that all the necessary preconditions for successful execution are met
+        # prior to execution. That means if a call fails, it may result in a modified
+        # and incomplete data structure that will be difficult to fix when the user is
+        # working interactively in the interpreter
+        if not Config.getOption('sbol_compliant_uris'):
+            raise EnvironmentError('Assemble method requires SBOL-compliance enabled')
+
+        # Validate doc
+        if not self.doc and not doc:
+            raise ValueError('Missing doc argument. If the ComponentDefinition does '
+                             'not belong to a Document, a target Document must be '
+                             'specified using the doc keyword argument.')
+        if doc and self.doc != doc:
+            raise ValueError('Invalid doc argument. Do not use the doc keyword '
+                             'argument if the ComponentDefinition already belongs '
+                             'to a Document')
+
+        # Validate component_list
+        doc = doc if doc else self.doc
+        if isinstance(component_list, list) and all(isinstance(c, ComponentDefinition)
+                                                    for c in component_list):
+            for cdef in component_list:
+                if cdef.doc and cdef.doc.this != doc.this:
```

Comment:
How does `cdef.doc.this != doc.this` differ from `cdef.doc != doc`? What does `this` do? `SBOLObject.this` is currently deprecated. Perhaps it shouldn't be? Or perhaps it is not implemented properly?

Reply:
This has been moved to #245